src/tree/updater_quantile_hist.cc

```diff
@@ -732,198 +770,237 @@ void QuantileHistMaker::Builder::EvaluateSplit(const std::vector<ExpandEntry>& n
     }
   }
 
-  builder_monitor_.Stop("EvaluateSplit");
+  builder_monitor_.Stop("EvaluateSplits");
 }
 
-void QuantileHistMaker::Builder::ApplySplit(int nid,
-                                            const GHistIndexMatrix& gmat,
-                                            const ColumnMatrix& column_matrix,
-                                            const HistCollection& hist,
-                                            const DMatrix& fmat,
-                                            RegTree* p_tree) {
-  builder_monitor_.Start("ApplySplit");
-  // TODO(hcho3): support feature sampling by levels
-
-  /* 1. Create child nodes */
-  NodeEntry& e = snode_[nid];
-  bst_float left_leaf_weight =
-      spliteval_->ComputeWeight(nid, e.best.left_sum) * param_.learning_rate;
-  bst_float right_leaf_weight =
-      spliteval_->ComputeWeight(nid, e.best.right_sum) * param_.learning_rate;
-  p_tree->ExpandNode(nid, e.best.SplitIndex(), e.best.split_value,
-                     e.best.DefaultLeft(), e.weight, left_leaf_weight,
-                     right_leaf_weight, e.best.loss_chg, e.stats.sum_hess);
-
-  /* 2. Categorize member rows */
-  const auto nthread = static_cast<bst_omp_uint>(this->nthread_);
-  row_split_tloc_.resize(nthread);
-  for (bst_omp_uint i = 0; i < nthread; ++i) {
-    row_split_tloc_[i].left.clear();
-    row_split_tloc_[i].right.clear();
-  }
-  const bool default_left = (*p_tree)[nid].DefaultLeft();
-  const bst_uint fid = (*p_tree)[nid].SplitIndex();
-  const bst_float split_pt = (*p_tree)[nid].SplitCond();
-  const uint32_t lower_bound = gmat.cut.Ptrs()[fid];
-  const uint32_t upper_bound = gmat.cut.Ptrs()[fid + 1];
-  int32_t split_cond = -1;
-  // convert floating-point split_pt into corresponding bin_id
-  // split_cond = -1 indicates that split_pt is less than all known cut points
-  CHECK_LT(upper_bound,
-           static_cast<uint32_t>(std::numeric_limits<int32_t>::max()));
-  for (uint32_t i = lower_bound; i < upper_bound; ++i) {
-    if (split_pt == gmat.cut.Values()[i]) {
-      split_cond = static_cast<int32_t>(i);
-    }
-  }
+template <bool default_left>
+inline std::pair<size_t, size_t> PartitionDenseKernel(
+    const size_t *rid, const uint32_t *idx,
+    const uint32_t offset, const int32_t split_cond,
+    const size_t istart, const size_t iend, size_t *p_left,
+    size_t *p_right) {
+  size_t ileft = 0;
+  size_t iright = 0;
 
-  const auto& rowset = row_set_collection_[nid];
+  const uint32_t max_val = std::numeric_limits<uint32_t>::max();
 
-  Column column = column_matrix.GetColumn(fid);
-  if (column.GetType() == xgboost::common::kDenseColumn) {
-    ApplySplitDenseData(rowset, gmat, &row_split_tloc_, column, split_cond,
-                        default_left);
-  } else {
-    ApplySplitSparseData(rowset, gmat, &row_split_tloc_, column, lower_bound,
-                         upper_bound, split_cond, default_left);
+  for (size_t i = istart; i < iend; i++) {
```

Comment:
This section of code is hard to read. Consider improving it with better variable names or comments. e.g. istart and iend of what? What is p_left?

Reply:
I know that it's not good explanation, but it's not my piece of code and I tried not to change this.
What I can do now to make it easier (without complete rewriting) - make names more clear and add comments.

Reply:
Added a commit with the changes